### PR TITLE
Keep chat history visible on page reload

### PR DESCRIPTION
**Pull request by baha-a**
Related to:
- #460 

Combined with:
- https://github.com/buerokratt/Chat-Widget/pull/103

### CODE REVIEW

File: DSL/Ruuter.public/DSL/POST/remove-chat-from-termination-queue.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+extractRequestData:
+  assign:
+    chatId: "${incoming.body.message.chatId}"
```
Comment (1AhmedYasser) [inline DSL/Ruuter.public/DSL/POST/remove-chat-from-termination-queue.yml:3]: Please remove `.message`, it should be `chatId: "${incoming.body.chatId}"` as you don't pass message from chat widget side

File: notification-server/src/server.js [hunk 2]
```diff
@@ -33,12 +33,32 @@ app.get("/sse/queue/:id", (req, res) => {
...
+
+app.post("/add-chat-to-termination-queue", async (req, res) => {
+  terminationQueue.set(req.body.chatId, req.body);
+  res.status(200).json({ response: 'enqueued successfully' });
+});
+
+app.post("/remove-chat-from-termination-queue", async (req, res) => {
+  const body = terminationQueue.get(req.body.chatId);
+
+  await fetch(`${process.env.RUUTER_URL}/end-chat`, {
```
Comment (1AhmedYasser) [inline notification-server/src/server.js:54]: This part should not end the chat, it should only remove from queue

File: notification-server/src/server.js [hunk 3]
```diff
@@ -33,12 +33,32 @@ app.get("/sse/queue/:id", (req, res) => {
...
+    method: 'POST',
+    body: JSON.stringify(body),
+    headers: { 'Content-Type': 'application/json' },
+  });
+  
+  terminationQueue.delete(req.body.chatId);
+  res.status(200).json({ response: 'dequeued successfully' });
 });
```
Comment (1AhmedYasser) [inline notification-server/src/server.js:63]: Another part should be added to regularly check for chats in termination queue and check their times, if it for example exceeded 5 seconds, then the chat should be ended and removed from termination queue

File: DSL/Ruuter.public/DSL/POST/add-chat-to-termination-queue.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    status: "${incoming.body.status}"
+  next: updateChatStatus
+
+updateChatStatus:
+  call: http.post
+  args:
+    url: "[#CHATBOT_RESQL]/set-termination-status"
+    body:
+      chatId: ${chatId}
+      scheduledForTerminated: ${new Date().toISOString()}
```
Comment (1AhmedYasser) [inline DSL/Ruuter.public/DSL/POST/add-chat-to-termination-queue.yml:16]: I noticed in database that this entry is added as `yyyy-MM-dd`, it should be date and time as well in order to help when checking the termination queue to remove chats from it

**Review by 1AhmedYasser — CHANGES_REQUESTED** [review]

**Comment by 1AhmedYasser** [comment]
remove from termination queue only works with the first refresh. After that it gets cancelled resulting in the chat being in the termination queue in the database endlessly

File: DSL/Ruuter.public/DSL/POST/add-chat-to-termination-queue.yml [hunk 5]
```diff
@@ -0,0 +1,34 @@
+extractRequestData:
+  assign:
+    chatId: "${incoming.body.message.chatId}"
+    event: "${incoming.body.message.event}"
+    authorRole: ${incoming.body.message.authorRole}
```
Comment (1AhmedYasser) [inline DSL/Ruuter.public/DSL/POST/add-chat-to-termination-queue.yml:5]: Please add double quotes `authorRole: "${incoming.body.message.authorRole}"`

**Review by 1AhmedYasser — CHANGES_REQUESTED** [review]

**Comment by 1AhmedYasser** [comment]
https://github.com/buerokratt/Buerokratt-Chatbot/assets/26207361/94b2cdfb-f7da-4007-ab44-f44884eba992

@baha-a  There is an Edge case that i found when i was checking the latest changes.

1. Start a chat and forward it to the back-office
2. Close the tab, and wait 5 seconds, the chat will be ended automatically
3. Now start a new chat and forward it to the back-office
4. Now refresh and then close the tab, the chat will not end after 5 seconds and when you open the widget again and then close the tab then the chat will be ended.

In conclusion, if you refresh and then close the tab, then the chat will not end, it only ends if you didn't refresh before closing the tab

**Review by 1AhmedYasser — CHANGES_REQUESTED** [review]
Please address the edge case i mentioned

**Comment by baha-a** [comment]
> Please address the edge case i mentioned

This edge case has been resolved. Please review it again at your earliest convenience.
